Refactor the generation of the metadata for linking

bjorn3 opened the pull request.

rust-highfive commented:
Some changes occured to rustc_codegen_cranelift

cc @bjorn3

rust-highfive commented:
r? @jackh726

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2766013297) [(plain)](https://github.com/rust-lang/rust/commit/c3df57e9084b91129a37107814343dfd4df02262/checks/2766013297/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0061]: this function takes 1 argument but 0 arguments were supplied
    --> src/librustdoc/clean/utils.rs:33:25
     |
33   |     for &cnum in cx.tcx.crates().iter() {
     |
note: associated function defined here
    --> /checkout/compiler/rustc_middle/src/query/mod.rs:1447:11
     |
     |
12   | / rustc_queries! {
13   | |     query trigger_delay_span_bug(key: DefId) -> () {
14   | |         desc { "trigger a delay span bug" }
...    |
...    |
1447 | |     query crates(_: ()) -> &'tcx [CrateNum] {
...    |
1714 | |     }
1715 | | }
     | |_- in this expansion of `rustc_query_append!`
     | |_- in this expansion of `rustc_query_append!`
     | 
    ::: /checkout/compiler/rustc_middle/src/ty/query/mod.rs:284:1
     |
284  |   rustc_query_append! { [define_callbacks!][<'tcx>] }
     |   --------------------------------------------------- in this macro invocation
help: expected the unit value `()`; create it with empty parentheses
     |
33   |     for &cnum in cx.tcx.crates(()).iter() {

error[E0061]: this function takes 1 argument but 0 arguments were supplied
    --> src/librustdoc/passes/collect_trait_impls.rs:33:25
     |
     |
33   |     for &cnum in cx.tcx.crates().iter() {
     |
note: associated function defined here
    --> /checkout/compiler/rustc_middle/src/query/mod.rs:1447:11
     |
     |
12   | / rustc_queries! {
13   | |     query trigger_delay_span_bug(key: DefId) -> () {
14   | |         desc { "trigger a delay span bug" }
...    |
...    |
1447 | |     query crates(_: ()) -> &'tcx [CrateNum] {
...    |
1714 | |     }
1715 | | }
     | |_- in this expansion of `rustc_query_append!`
     | |_- in this expansion of `rustc_query_append!`
     | 
    ::: /checkout/compiler/rustc_middle/src/ty/query/mod.rs:284:1
     |
284  |   rustc_query_append! { [define_callbacks!][<'tcx>] }
     |   --------------------------------------------------- in this macro invocation
help: expected the unit value `()`; create it with empty parentheses
     |
33   |     for &cnum in cx.tcx.crates(()).iter() {

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0061`.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2766061743) [(plain)](https://github.com/rust-lang/rust/commit/563ed8c0bb3445b54a14810083f24cab3022ea6a/checks/2766061743/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_middle/src/middle/cstore.rs at line 210:
 // In order to get this left-to-right dependency ordering, we use the reverse
 // postorder of all crates putting the leaves at the right-most positions.
 pub fn used_crates(tcx: TyCtxt<'_>) -> Vec<CrateNum> {
-    tcx
-        .postorder_cnums(())
+    tcx.postorder_cnums(())
         .iter()
         .rev()
         .cloned()
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_middle/src/mir/visit.rs" "/checkout/compiler/rustc_middle/src/mir/coverage.rs" "/checkout/compiler/rustc_middle/src/middle/cstore.rs" "/checkout/compiler/rustc_middle/src/mir/abstract_const.rs" "/checkout/compiler/rustc_middle/src/middle/exported_symbols.rs" "/checkout/compiler/rustc_middle/src/mir/traversal.rs" "/checkout/compiler/rustc_middle/src/middle/privacy.rs" "/checkout/compiler/rustc_middle/src/middle/region.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:18
```

</details>

bjorn3 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit fa21dd12e6bd6c74cb0b96e94933212239944a0a with merge 15f098a756314fabc76b01fc133b3e42a6d4b740...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fa21dd12e6bd6c74cb0b96e94933212239944a0a","merge_sha":"15f098a756314fabc76b01fc133b3e42a6d4b740"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2766727196)
Build commit: 15f098a756314fabc76b01fc133b3e42a6d4b740 (`15f098a756314fabc76b01fc133b3e42a6d4b740`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2766727196"},"merge_sha":"15f098a756314fabc76b01fc133b3e42a6d4b740"} -->

rust-timer commented:
Queued 15f098a756314fabc76b01fc133b3e42a6d4b740 with parent 2312ff1a850db449b79fd3c4b215395cd2598c25, future [comparison URL](https://perf.rust-lang.org/compare.html?start=2312ff1a850db449b79fd3c4b215395cd2598c25&end=15f098a756314fabc76b01fc133b3e42a6d4b740).

rust-timer commented:
Finished benchmarking try commit (15f098a756314fabc76b01fc133b3e42a6d4b740): [comparison url](https://perf.rust-lang.org/compare.html?start=2312ff1a850db449b79fd3c4b215395cd2598c25&end=15f098a756314fabc76b01fc133b3e42a6d4b740).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -256,22 +256,22 @@ pub fn each_linked_rlib(
-            LibSource::None => return Err(format!("could not find rlib for: `{}`", name)),
+        let path = if let Some((path, _)) = &info.used_crate_source[&cnum].rlib {
+            path
+        } else if info.used_crate_source[&cnum].rmeta.is_some() {
```
The `info.used_crate_source[&cnum]` search can be done once instead of twice here.

petrochenkov commented:
All changes look good to me, but the source of systematic regressions in https://github.com/rust-lang/rust/pull/86105#issuecomment-856243484 needs to be figured out.

You can break this PR into parts for that and assign them to me, I'll r+ them quickly if they don't cause perf regressions.

bjorn3 commented:
Rebased on top of  #86877 and removed 
279acba as I committed it directly to cg_clif's repo in a slightly changed form: https://github.com/bjorn3/rustc_codegen_cranelift/commit/fed71e344818dd2b1f2e48005b6d3da21cda30f5

bjorn3 commented:
I will do a final perf run on this PR. If it is a regression, I will close it.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 870a559178e46ba1ced1f90854a0839651a7a235 with merge 9960b2de2485491f52fa389367b9c5faea239d7d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"870a559178e46ba1ced1f90854a0839651a7a235","merge_sha":"9960b2de2485491f52fa389367b9c5faea239d7d"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2992419075)
Build commit: 9960b2de2485491f52fa389367b9c5faea239d7d (`9960b2de2485491f52fa389367b9c5faea239d7d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2992419075"},"merge_sha":"9960b2de2485491f52fa389367b9c5faea239d7d"} -->

rust-timer commented:
Queued 9960b2de2485491f52fa389367b9c5faea239d7d with parent 6e9b3696d494a32d493585f96f0671123066cd58, future [comparison URL](https://perf.rust-lang.org/compare.html?start=6e9b3696d494a32d493585f96f0671123066cd58&end=9960b2de2485491f52fa389367b9c5faea239d7d).

rust-timer commented:
Finished benchmarking try commit (9960b2de2485491f52fa389367b9c5faea239d7d): [comparison url](https://perf.rust-lang.org/compare.html?start=6e9b3696d494a32d493585f96f0671123066cd58&end=9960b2de2485491f52fa389367b9c5faea239d7d).

**Summary**: This benchmark run did not return any significant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up.  

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

bjorn3 commented:
Perf is much better this time! There are still regressions, but they are smaller and there are a couple of improvements up to 0.7%.

petrochenkov commented on compiler/rustc_middle/src/middle/cstore.rs:
```diff
@@ -204,23 +204,18 @@ pub type CrateStoreDyn = dyn CrateStore + sync::Sync;
+    tcx.postorder_cnums(())
         .iter()
+        .rev()
         .cloned()
```
```suggestion
        .copied()
```

petrochenkov commented on compiler/rustc_middle/src/middle/cstore.rs:
```diff
@@ -204,23 +204,18 @@ pub type CrateStoreDyn = dyn CrateStore + sync::Sync;
-    ordering.reverse();
-    libs.sort_by_cached_key(|&a| ordering.iter().position(|x| *x == a));
-    libs
+        .collect::<Vec<_>>()
```
```suggestion
        .collect()
```

petrochenkov commented:
I have no idea why this caused performance differences.
The produced result should be the same, and from code changes it also should be an optimization.

petrochenkov commented:
r=me with the nits addressed.

bjorn3 commented:
Fixed the nits and also switched from `filter_map` to plain `filter` as nothing was mapped anyway.

petrochenkov commented:
The function can probably be inlined at this point, it's used only once anyway.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3001601422) [(plain)](https://github.com/rust-lang/rust/commit/b7d6da6b12ba4f0cb3f0fa4216eaa69a95240690/checks/3001601422/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0308]: mismatched types
   --> compiler/rustc_middle/src/middle/cstore.rs:214:38
    |
214 |         .filter(|cnum| !tcx.dep_kind(cnum).macros_only())
    |                                      |
    |                                      |
    |                                      expected struct `rustc_span::def_id::CrateNum`, found `&rustc_span::def_id::CrateNum`
    |                                      help: consider dereferencing the borrow: `*cnum`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rustc_middle`
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3001695789) [(plain)](https://github.com/rust-lang/rust/commit/4cd980841f448f469a9c21548d8bc693a06eeb9a/checks/3001695789/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
error: unused import: `self`
  --> compiler/rustc_codegen_ssa/src/base.rs:22:36
   |
22 | use rustc_middle::middle::cstore::{self, EncodedMetadata};
   |
   |
   = note: `-D unused-imports` implied by `-D warnings`
    Checking rustc_resolve v0.0.0 (/checkout/compiler/rustc_resolve)
error: aborting due to previous error

error: could not compile `rustc_codegen_ssa`
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3008803201) [(plain)](https://github.com/rust-lang/rust/commit/2f52e1cad668f386b2bf1c054c875a5b92d3a4eb/checks/3008803201/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_codegen_ssa/src/base.rs at line 783:
         //
         // In order to get this left-to-right dependency ordering, we use the reverse
         // postorder of all crates putting the leaves at the right-most positions.
-        let used_crates = tcx.postorder_cnums(())
+        let used_crates = tcx
+            .postorder_cnums(())
             .iter()
             .rev()
             .copied()
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_codegen_ssa/src/back/write.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/write.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/builder.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/consts.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/rpath.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/linker.rs" "/checkout/compiler/rustc_codegen_ssa/src/base.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/debuginfo.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

petrochenkov commented:
r=me with squashed commits and green CI.

bjorn3 commented:
Splat

@bors r=petrochenkov

(CI was green when I squashed. I didn't rebase, so the tree is identical.)

bors commented:
:pushpin: Commit 2977dff116a3ec93135c349ed39a46eec008c51c has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 2977dff116a3ec93135c349ed39a46eec008c51c -->
<!-- homu: {"type":"Approved","sha":"2977dff116a3ec93135c349ed39a46eec008c51c","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 2977dff116a3ec93135c349ed39a46eec008c51c with merge 5150aa50926afd13216225559fc0e26dcdb7aa51...
<!-- homu: {"type":"BuildStarted","head_sha":"2977dff116a3ec93135c349ed39a46eec008c51c","merge_sha":"5150aa50926afd13216225559fc0e26dcdb7aa51"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3010292971)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/3010292971","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3010292971) [(plain)](https://github.com/rust-lang-ci/rust/commit/5150aa50926afd13216225559fc0e26dcdb7aa51/checks/3010292971/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bjorn3 commented:
The dist-aarch64-apple job seems to not have started. This is the full log:

```
2021-07-07T13:35:42.5693605Z Found online and idle hosted runner(s) in the current repository's organization account that matches the required labels: 'macos-latest'
2021-07-07T13:35:42.5693641Z Waiting for a hosted runner in 'organization' to pick this job...
```

@bors retry

bors commented:
:hourglass: Testing commit 2977dff116a3ec93135c349ed39a46eec008c51c with merge d2b04f075c0ce010758c4c8674152ff89d1d73f3...
<!-- homu: {"type":"BuildStarted","head_sha":"2977dff116a3ec93135c349ed39a46eec008c51c","merge_sha":"d2b04f075c0ce010758c4c8674152ff89d1d73f3"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3013156081)
Approved by: petrochenkov
Pushing d2b04f075c0ce010758c4c8674152ff89d1d73f3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3013156081"},"merge_sha":"d2b04f075c0ce010758c4c8674152ff89d1d73f3"} -->
